### PR TITLE
PV Controller: Add plugin name and volume mode to PV metrics

### PR DESCRIPTION
**Pull request by tsmetana**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
The PV Controller contains pv_collector_(un)bound_pv_count metrics that count nubmer of PVs by the StorageName. The StorageName is however not that important information when monitoring larger number of clusters since it doesn't really determine the type of storage. The PR makes this metric more informative by adding also the plugin name and volume mode to the metric.

**Does this PR introduce a user-facing change?**:

```release-note
There is a new pv_collector_total_pv_count metric that counts persistent volumes by the volume plugin name and volume mode.
```

### CODE REVIEW

File: pkg/controller/volume/persistentvolume/metrics/metrics.go [hunk 1]
```diff
@@ -110,6 +118,23 @@ var (
 		[]string{"plugin_name", "operation_name"})
 )
 
+// volumeCount counts by PluginName, StorageClass and VolumeMode.
+type volumeCount map[string]map[string]map[string]int
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/metrics/metrics.go:122]: Can we improve readability of this structs, we can do perhaps define named type (like type pluginName string) and then use it in aggregate type.

File: pkg/controller/volume/persistentvolume/metrics/metrics.go [hunk 2]
```diff
@@ -122,33 +147,55 @@ func (collector *pvAndPVCCountCollector) CollectWithStability(ch chan<- metrics.
...
+			for volumeMode, number := range volumeModeCount {
+				ch <- metrics.NewLazyConstMetric(
+					boundPVCountDesc,
+					metrics.GaugeValue,
+					float64(number),
+					storageClassName,
+					pluginName,
+					volumeMode)
+			}
+		}
```
Comment (gnufied) [inline pkg/controller/volume/persistentvolume/metrics/metrics.go:185]: some new unit tests for checking these labels would be nice.

**Comment by gnufied** [comment]
Mostly looks good to me. cc @msau42 what do you think about adding new labels to these metrics? since these metrics are marked as alpha, it should be fine but we might break some existing metric dashboards.

**Comment by tsmetana** [comment]
We talked with @gnufied and @jsafrane and agreed that it might be safer to add a new pv_collector metric without introducing incompatible changes to the existing metrics (albeit alpha).

**Comment by tsmetana** [comment]
@gnufied here's the new metric. I'm now trying to update the e2e: it's quite tailored for the one-dimensional metrics... Still, I would appreciate if you would tell me if this is it.

**Comment by tsmetana** [comment]
I added an e2e test (separate commit) that should verify the new metrics gets created and has the required labels.

File: pkg/controller/volume/persistentvolume/metrics/metrics.go (hunk 1 above)
Comment (tsmetana) [reply pkg/controller/volume/persistentvolume/metrics/metrics.go]: It's simpler now and looks completely like the metrics in the volume manager. So I hope it's enough.

File: pkg/controller/volume/persistentvolume/metrics/metrics.go (hunk 2 above)
Comment (tsmetana) [reply pkg/controller/volume/persistentvolume/metrics/metrics.go]: There are some e2e tests now added to the existing ones.

**Comment by tsmetana** [comment]
@gnufied @msau42 I've updated the patch, PTAL.

**Comment by dashpole** [comment]
/assign @logicalhan 
for instrumentation

**Comment by gnufied** [comment]
from storage side lgtm.

/lgtm

**Comment by jsafrane** [comment]
/approve

Should the release note say `Breaking change:  the alpha metrics X and Y ....`?

**Comment by gnufied** [comment]
actually release note should be updated. we are no longer modifying the existing metrics.

File: pkg/controller/volume/persistentvolume/metrics/metrics.go [hunk 3]
```diff
@@ -73,12 +80,19 @@ type pvAndPVCCountCollector struct {
...
 	pvcLister PVCLister
+	// Volume plugin manager
+	pluginMgr *volume.VolumePluginMgr
 }
 
 // Check if our collector implements necessary collector interface
 var _ metrics.StableCollector = &pvAndPVCCountCollector{}
 
 var (
+	totalPVCountDesc = metrics.NewDesc(
```
Comment (msau42) [inline pkg/controller/volume/persistentvolume/metrics/metrics.go:91]: What do you think of adding the phase and datasource kind as a metric label?

**Review by msau42 — COMMENTED** [review]
cc @mattcary

**Comment by jsafrane** [comment]
/hold cancel
I can see the release note is OK.

File: pkg/controller/volume/persistentvolume/metrics/metrics.go (hunk 3 above)
Comment (tsmetana) [reply pkg/controller/volume/persistentvolume/metrics/metrics.go:91]: Hm... I actually don't have any opinion on that. :) But if we agree it's useful to add them here, I don't mind doing it. These metrics are still alpha, I assume it's also for us to figure out which are useful and give us opportunity to make changes.

Comment (msau42) [reply pkg/controller/volume/persistentvolume/metrics/metrics.go:91]: I think they would be useful so you can also get a count of pvs in bound/available/released/lost phases, and datasource is useful so you can tell if it's a snapshot restore/clone/populator/etc

**Pull request merged**